extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -68,121 +86,141 @@ export class Messaging {
+  sendMessage_(mId, mType, mName, mData, mRsvp, mErr) {
+    /** @type {Message} */
     const message = {
-      sentinel,
-      requestId,
-      type: eventType,
-      payload,
-      rsvp: awaitResponse,
+      app: APP,
+      requestid: mId,
+      type: mType,
+      name: mName,
+      data: mData,
+      rsvp: mRsvp,
+      error: mErr,
     };
     this.target_./*OK*/postMessage(message, this.targetOrigin_);
   }
 
   /**
-   * @param {string} requestId
+   * @param {number} requestId
+   * @param {string} name
    * @param {*} reason
    * @private
    */
-  sendResponseError_(requestId, reason) {
+  sendResponseError_(requestId, name, reason) {
     this.sendMessage_(
-      responseSentinel_, requestId, 'ERROR', reason, false);
+      requestId, MessageType_.RESPONSE, name, null, null, reason);
   }
 
   /**
    * I'm handing an incoming request from Bob. I'll either respond normally
    * (ex: "got it Bob!") or with an error (ex: "I didn't get a word of what
    * you said!").
-   * @param {*} message
+   * @param {Message} message
    * @private
    */
   handleRequest_(message) {
-    dev().info(TAG, 'messaging.js -> handleRequest_');
-    const requestId = message.requestId;
-    const promise = this.requestProcessor_(message.type, message.payload,
-        message.rsvp);
+    dev().info(TAG, 'messaging.js -> handleRequest_', message);
+    dev().assert(this.requestProcessor_,
+      'Cannot handle request because handshake is not yet confirmed!');
+    const requestId = message.requestid;
+    const msg = dev().assertString(message.name);
     if (message.rsvp) {
+      const promise =
+        this.requestProcessor_(msg, message.data, message.rsvp);
       if (!promise) {
-        this.sendResponseError_(requestId, 'no response');
+        this.sendResponseError_(requestId, msg, 'no response');
         dev().assert(promise,
-          'expected response but none given: ' + message.type);
+          'expected response but none given: ' + message.name);
       }
-      promise.then(payload => {
-        this.sendResponse_(requestId, payload);
+      promise.then(data => {
+        this.sendResponse_(requestId, msg, data);
       }, reason => {
-        this.sendResponseError_(requestId, reason);
+        this.sendResponseError_(requestId, msg, reason);
       });
     }
   }
 
   /**
    * I sent out a request to Bob. He responded. And now I'm handling that
    * response.
-   * @param {*} message
+   * @param {Message} message
    * @private
    */
   handleResponse_(message) {
     dev().info(TAG, 'messaging.js -> handleResponse_');
-    const requestId = message.requestId;
+    const requestId = message.requestid;
     const pending = this.waitingForResponse_[requestId];
     if (pending) {
       delete this.waitingForResponse_[requestId];
-      if (message.type == 'ERROR') {
-        pending.reject(message.payload);
+      if (message.error) {
+        pending.reject(/** @type {!Error} */ (message.error));
```

Comment:
Is this a correct cast? Is it even possible to transfer an Error object via messaging?